### PR TITLE
[JAVA][#5172] Allow vendor json media types

### LINKED ISSUE #5172: [JAVA] isJsonMime should support vendor specific JSON MIME type
##### Description

We support APIs that return JSON encoded responses with vendor specific mime types.  These MIME types look something like:
```
application/vnd.company+json
```
Which is a valid MIME-type. A response with the above mime-type as its content-type indicates that the response is JSON encoded.  When the swagger-client attempts to deserialize such response, it throws an exception due to a method responsible for determining whether the given MIME is a JSON MIME.  The method is [`isJsonMime`](https://github.com/swagger-api/swagger-codegen/blob/3c1a8daaf920af7a5eecf17f3fb157f083ca6366/modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/ApiClient.mustache#L728) and it checks whether the mime type matches the following regex:
```
(?i)application\\/json(;.*)?
```

This results in the following error:
```
Content type "application/vnd.company+json" is not supported for type: class com.somemodel Model
```

Not tested, but I think the above regex should be updated to something like:

```
(?i)application\\/((\w\.)*\w\+)?json(;.*)?"
```
so that MIME types like `application/vnd.company+json` are recognized as a valid JSON MIME.

I think it would be valid to have some additional tests [here](https://github.com/swagger-api/swagger-codegen/blob/1dd9ee39af6c4e2f9c016a94d16ca47a88a9cb0d/samples/client/petstore/java/okhttp-gson/src/test/java/io/swagger/client/ApiClientTest.java#L88) 
that checks for MIME types such as the one mentioned above.  

##### Swagger-codegen version

`v2.2.1`

##### Swagger declaration file content or url

N/A

##### Command line used for generation

```bash
java -jar /opt/swagger-codegen/modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate -i /mnt/swagger.yaml -l java -o /mnt/clients/java -c /mnt/clients/configs/java.json
```

##### Steps to reproduce

- Generate a client for some API
- Have the API return a vendor JSON MIME type as the content header
- use the client to retrieve from the endpoint which produces a vendor JSON MIME type
- watch it scream

##### Related issues

This is a simpler fix, but I believe https://github.com/swagger-api/swagger-codegen/issues/3870 is related.

##### Suggest a Fix

Mentioned in the description.

### PR DESCRIPTION
**Pull request by davidharrigan**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [x] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR
Adds support for vendor media type JSON (ends with `+json`) as specified in [RFC-3839](https://tools.ietf.org/html/rfc6839#section-3.1). 

PR made against master since the change is non-breaking and backwards compatible (still has the same support for `application/json`). 

More details in the issue here: #5172

### CODE REVIEW

**Comment by davidharrigan** [comment]
Not sure whether I'm doing something wrong, but this test fails for me on master after I rebuild the petstore sample using the provided scripts in the bin directory.
```bash
┌─dharrigan•mbpr petstore/java/jersey2
└─→  mvn verify                                                                                                                master :: 1d :: ⬢
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for io.swagger:swagger-petstore-jersey2:jar:1.0.0
[WARNING] 'build.plugins.plugin.version' for org.codehaus.mojo:build-helper-maven-plugin is missing. @ line 87, column 15
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building swagger-petstore-jersey2 1.0.0
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- build-helper-maven-plugin:3.0.0:add-source (add_sources) @ swagger-petstore-jersey2 ---
[INFO] Source directory: /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/src/main/java added.
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ swagger-petstore-jersey2 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:2.5.1:compile (default-compile) @ swagger-petstore-jersey2 ---
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 47 source files to /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/target/classes
[INFO]
[INFO] --- build-helper-maven-plugin:3.0.0:add-test-source (add_test_sources) @ swagger-petstore-jersey2 ---
[INFO] Test Source directory: /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/src/test/java added.
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ swagger-petstore-jersey2 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:2.5.1:testCompile (default-testCompile) @ swagger-petstore-jersey2 ---
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 1 source file to /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:2.12:test (default-test) @ swagger-petstore-jersey2 ---
[INFO] Surefire report directory: /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.api.FakeApiTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.api.PetApiTest
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.007 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.api.StoreApiTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.115 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.api.UserApiTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.142 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.ApiClientTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.339 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.auth.ApiKeyAuthTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.auth.HttpBasicAuthTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.ConfigurationTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.329 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.JSONTest
Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.436 sec <<< FAILURE!
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.model.EnumValueTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.28 sec
Concurrency config is parallel='methods', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running io.swagger.client.StringUtilTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec

Results :

Failed tests:   testSqlDateSerialization(io.swagger.client.JSONTest): expected:<"19[70-01-0]1"> but was:<"19[69-12-3]1">

Tests run: 42, Failures: 1, Errors: 0, Skipped: 1

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13.424 s
[INFO] Finished at: 2017-03-24T12:11:37-04:00
[INFO] Final Memory: 22M/308M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12:test (default-test) on project swagger-petstore-jersey2: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/dharrigan/work/src/swagger-codegen/samples/client/petstore/java/jersey2/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

File: modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache [hunk 1]
```diff
@@ -458,11 +458,13 @@ public class ApiClient {
    *   application/json
    *   application/json; charset=UTF8
    *   APPLICATION/JSON
+   *   application/vnd.company+json
    * @param mime MIME
    * @return True if MIME type is boolean
    */
   public boolean isJsonMime(String mime) {
-    return mime != null && mime.matches("(?i)application\\/json(;.*)?");
+    String jsonMime = "(?i)^(application/json|[^;/ \t]+/[^;/ \t]+[+]json)[ \t]*(;.*)?$";
```
Comment (ePaul) [inline modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache]: Are tab characters really allowed in media type names?

**Comment by ePaul** [comment]
Could you rebase your branch onto current master, and possibly regenerate the samples afterwards?

I think some of the changes in the generated samples are already incorporated there, and this would make it easier to see what actually changed.

File: samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java [hunk 2]
```diff
@@ -40,15 +40,24 @@ public void testParseAndFormatDate() {
...
+
+        assertTrue(apiClient.isJsonMime("application/json"));
+        assertTrue(apiClient.isJsonMime("application/json; charset=UTF8"));
+        assertTrue(apiClient.isJsonMime("APPLICATION/JSON"));
+
+        assertTrue(apiClient.isJsonMime("application/json\t"));
+        assertTrue(apiClient.isJsonMime("example/foo+bar+json"));
+        assertTrue(apiClient.isJsonMime("example/foo+json;x;y"));
+        assertTrue(apiClient.isJsonMime("example/foo+json\t;"));
+        assertTrue(apiClient.isJsonMime("Example/fOO+JSON"));
```
Comment (ePaul) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: If you want a real-live example, `application/problem+json` (from [RFC 7807](https://tools.ietf.org/html/rfc7807)) is often used in APIs.

File: modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache (hunk 1 above)
Comment (davidharrigan) [inline modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache]: Yes, according to the grammar described here in [RFC-7230](https://tools.ietf.org/html/rfc7230#section-3.2).  You can see that `HTAB` corresponds to horizontal tab here in [RFC-5234](https://tools.ietf.org/html/rfc5234#appendix-B.1).

File: samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java (hunk 2 above)
Comment (davidharrigan) [inline samples/client/petstore/java/jersey1/src/test/java/io/swagger/client/ApiClientTest.java]: Sure thing, I'll update these tests with some real-live example.

File: modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache (hunk 1 above)
Comment (ePaul) [inline modules/swagger-codegen/src/main/resources/Java/ApiClient.mustache]: Okay, thanks.

**Comment by davidharrigan** [comment]
Looks like the failed test is fixed here: https://github.com/swagger-api/swagger-codegen/pull/5243
```
Failed tests:   testApiClient(io.swagger.client.api.PetApiTest): expected:<.../petstore.swagger.io[]/v2> but was:<.../petstore.swagger.io[:80]/v2>
```

**Comment by ePaul** [comment]
Yes, sorry for the update proposal, I didn't mean to cause more issues by that, but simply forgot that current master is "broken" (for sample generation) until #5243 is merged.

**Comment by davidharrigan** [comment]
no worries :)

File: modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/ApiClient.mustache [hunk 3]
```diff
@@ -720,12 +720,13 @@ public class ApiClient {
...
      *   APPLICATION/JSON
-     *
+     *   application/vnd.company+json
      * @param mime MIME (Multipurpose Internet Mail Extensions)
      * @return True if the given MIME is JSON, false otherwise.
      */
     public boolean isJsonMime(String mime) {
-        return mime != null && mime.matches("(?i)application\\/json(;.*)?");
+      String jsonMime = "(?i)^(application/json|[^;/ \t]+/[^;/ \t]+[+]json)[ \t]*(;.*)?$";
+      return mime != null && mime.matches(jsonMime);
```
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/ApiClient.mustache]: Later I will file a PR to put these 2 lines into one with declaring a temporary variable.

File: samples/client/petstore/java/okhttp-gson/src/test/java/io/swagger/client/ApiClientTest.java [hunk 4]
```diff
@@ -79,15 +79,26 @@ public void testParseAndFormatDate() {
...
+        assertTrue(apiClient.isJsonMime("application/json; charset=UTF8"));
+        assertTrue(apiClient.isJsonMime("APPLICATION/JSON"));
+
+        assertTrue(apiClient.isJsonMime("application/problem+json"));
+        assertTrue(apiClient.isJsonMime("APPLICATION/PROBLEM+JSON"));
+        assertTrue(apiClient.isJsonMime("application/json\t"));
+        assertTrue(apiClient.isJsonMime("example/foo+bar+json"));
+        assertTrue(apiClient.isJsonMime("example/foo+json;x;y"));
+        assertTrue(apiClient.isJsonMime("example/foo+json\t;"));
+        assertTrue(apiClient.isJsonMime("Example/fOO+JSON"));
```
Comment (wing328) [inline samples/client/petstore/java/okhttp-gson/src/test/java/io/swagger/client/ApiClientTest.java]: We'll leverage these test cases in other API clients' unit tests

**Comment by wing328** [comment]
@ePaul thanks for reviewing the change.

@davidharrigan thanks for the contribution, which has been merged into master.

**Pull request merged**